fate/src/main/java/org/apache/accumulo/fate/util/LoggingRunnable.java

```diff
@@ -34,20 +34,37 @@ public void run() {
     try {
       runnable.run();
     } catch (Throwable t) {
+      boolean runnableError = (t instanceof Error);
       try {
         log.error("Thread \"{}\" died {}", Thread.currentThread().getName(), t.getMessage(), t);
       } catch (Throwable t2) {
-        // maybe the logging system is screwed up OR there is a bug in the exception, like
-        // t.getMessage() throws a NPE
-        System.err.println(
-            "ERROR " + new Date() + " Failed to log message about thread death " + t2.getMessage());
-        t2.printStackTrace();
-
-        // try to print original exception
-        System.err
-            .println("ERROR " + new Date() + " Exception that failed to log : " + t.getMessage());
-        t.printStackTrace();
+        boolean logError = (t2 instanceof Error);
+        try {
+          // maybe the logging system is screwed up OR there is a bug in the exception, like
+          // t.getMessage() throws a NPE
+          System.err.println("ERROR " + new Date() + " Failed to log message about thread death "
+              + t2.getMessage());
+          t2.printStackTrace();
+
+          // try to print original exception
+          System.err
+              .println("ERROR " + new Date() + " Exception that failed to log : " + t.getMessage());
+          t.printStackTrace();
+        } catch (Throwable t3) {
+          // If printing to System.err didn't work then don't try to log that exception but do
+          // re-throw it if
+          // it's the most serious failure we've seen so far.
+          boolean printError = (t3 instanceof Error);
+          if (printError && !(logError || runnableError))
+            throw t3;
+        }
+
+        // If we got a more serious failure when attempting to log the message,
+        // then throw that instead of the original exception.
+        if (logError && !runnableError)
+          throw t2;
```

Comment:
Instead of choosing which one to throw, why not add the more recent error as a suppressed exception to the original one and still throw the original one?

Reply:
> Instead of choosing which one to throw, why not add the more recent error as a suppressed exception to the original one and still throw the original one?

I initially thought about combining the exceptions, but then I didn't because I was concerned that if we got here because of an OutOfMemoryError then performing an operation that requires a memory allocation (a new ArrayList in the case of adding the suppressed exception) would be likely to fail. Maybe that's not really an issue, since in either case, we'd throw an OOME. Also, this was intended to handle the case where we received an exception that falls under the Exception tree, but in trying to log, we had a failure that falls under the Error tree. In that case, we would still throw an Exception with a suppressed Error included. Would that break any VM handling such as killing on OOME, if configured? If not, then I think what you are suggesting would make the code more readable.

Reply:
@ctubbsii, I saw you approved but just wanted to double-check (since my comment above may have lost context when I pushed my last changes). Are you ok with me committing this as-is, or are there still changes you would like here?

Reply:
It's fine as-is. My idea was only a suggestion. I can see pros and cons with it.